rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/UserResource.java

```diff
@@ -0,0 +1,68 @@
+package org.keycloak.admin.ui.rest;
+
+import static org.keycloak.userprofile.UserProfileContext.USER_API;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+import org.jboss.resteasy.annotations.cache.NoCache;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.representations.userprofile.config.UPConfig;
+import org.keycloak.userprofile.UserProfile;
+import org.keycloak.userprofile.UserProfileProvider;
+
+/**
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class UserResource {
+
+    private final KeycloakSession session;
+    private final UserModel user;
+
+    public UserResource(KeycloakSession session, UserModel user) {
+        this.session = session;
+        this.user = user;
+    }
+
+    @GET
+    @Path("unmanagedAttributes")
+    @NoCache
+    @Produces(MediaType.APPLICATION_JSON)
+    public Map<String, List<String>> getUnmanagedAttributes() {
+        RealmModel realm = session.getContext().getRealm();
+        UserProfileProvider provider = session.getProvider(UserProfileProvider.class);
+
+        if (provider.isEnabled(realm)) {
+            UserProfile profile = provider.create(USER_API, user);
+            Map<String, List<String>> managedAttributes = profile.getAttributes().getReadable(false);
+            Map<String, List<String>> attributes = new HashMap<>(user.getAttributes());
+            UPConfig upConfig = provider.getConfiguration();
+
+            if (upConfig.getUnmanagedAttributePolicy() == null) {
+                return Collections.emptyMap();
+            }
+
+            Map<String, List<String>> unmanagedAttributes = profile.getAttributes().getUnmanagedAttributes();
+            managedAttributes.entrySet().removeAll(unmanagedAttributes.entrySet());
+            attributes.entrySet().removeAll(managedAttributes.entrySet());
+
+            attributes.remove(UserModel.USERNAME);
+            attributes.remove(UserModel.EMAIL);
+            attributes.remove(UserModel.FIRST_NAME);
```

Comment:
That is a good question and perhaps we should actually enforce these attributes are never removed ....

I also ended up in a weird situation when testing this PR locally where I was able to clear all attributes from the user profile configuration but I was still able to manage both username and email. The first name and last name were indeed removed. And to make things a bit worse, I can not create the `firstName` as an unmanaged attribute ...

Can we create a follow-up issue for this? It will probably be my next one so that we make sure the root attributes (username, email, firstname, and lastnames) are never removed.

Reply:
Hmm... That is good question.

I suppose that we can make sure to avoid removing username and email as those are "core" attributes used for authentication etc and also there are lots of corner cases for these (all those switches like `emailAsUsername` etc).

For firstName and lastname, it can be nice if it's possible to treat them as normal attributes (even if enabled by default for backwards compatibility etc) and have possibility to remove them? Or at least have the option to hide them from the forms (so not readable and writable for usernames and admins)?

Reply:
+1 as a follow-up. I am approving this PR, so we have basics around migration available.

Reply:
@mposolda Created https://github.com/keycloak/keycloak/issues/25147.